pyartifactory/objects/repository.py

```diff
@@ -35,42 +33,48 @@ def get_repo(self, repo_name: str) -> AnyRepositoryResponse:
         """
         Finds repository in artifactory. Raises an exception if the repo doesn't exist.
         :param repo_name: Name of the repository to retrieve
-        :return: Either a local, virtual or remote repository
+        :return: Either a local, virtual, remote or federated repository
         """
         try:
             response = self._get(f"api/{self._uri}/{repo_name}")
+            response_data = response.json()
+            rclass = None
+
             try:
-                artifact_info: AnyRepositoryResponse = LocalRepositoryResponse.model_validate(response.json())
-            except ValidationError:
-                try:
-                    artifact_info = VirtualRepositoryResponse.model_validate(response.json())
-                except ValidationError:
-                    artifact_info = RemoteRepositoryResponse.model_validate(response.json())
-            return artifact_info
+                rclass = response_data["rclass"]
+            except KeyError:
+                raise KeyError('"rclass" key not found in the response data received by artifactory.')
+
+            # Match to the correct repository type depending on the rclass
+            if rclass == RClassEnum.local:
+                return LocalRepositoryResponse.model_validate(response_data)
+            elif rclass == RClassEnum.virtual:
+                return VirtualRepositoryResponse.model_validate(response_data)
+            elif rclass == RClassEnum.remote:
+                return RemoteRepositoryResponse.model_validate(response_data)
+            elif rclass == RClassEnum.federated:
+                return FederatedRepositoryResponse.model_validate(response_data)
+            else:
+                # this should never happen and is a missing repotype in the library
+                raise ArtifactoryError(
+                    f"Unknown repository type found in response: {rclass}. Please report this issue.",
+                )
         except requests.exceptions.HTTPError as error:
             http_response: Union[Response, None] = error.response
             if isinstance(http_response, Response) and http_response.status_code in (404, 400):
                 logger.error("Repository %s does not exist", repo_name)
                 raise RepositoryNotFoundError(f" Repository {repo_name} does not exist")
             raise ArtifactoryError from error
 
-    @overload
-    def create_repo(self, repo: LocalRepository) -> LocalRepositoryResponse:
-        ...
-
-    @overload
-    def create_repo(self, repo: VirtualRepository) -> VirtualRepositoryResponse:
-        ...
-
-    @overload
-    def create_repo(self, repo: RemoteRepository) -> RemoteRepositoryResponse:
-        ...
```

Comment:
Why did you remove that ? It is important for the typing. Please revert it and add a section for federated repositories

Reply:
are you shure that it is nessecary?
1. it simply breaks mypy because local and feaderated signatures are too similar as mypy could handle this?
2. i THINK (with my limited knowledge of python typing) these where simply not nessecary because the typing is handled by the `AnyRepository` in models/__init__.py?

Reply:
the error when simply adding a section for the federated repo is:
```
pyartifactory/objects/repository.py:74: error: Overloaded function signatures 1
and 2 overlap with incompatible return types  [misc]
        def create_repo(self, repo: FederatedRepository) -> FederatedRepos...
        ^
pyartifactory/objects/repository.py:115: error: Overloaded function signatures
1 and 2 overlap with incompatible return types  [misc]
        def update_repo(self, repo: FederatedRepository) -> FederatedRepos...
        ^
```

Reply:
when looking at the code, I'm not shure if "reducing the code dupication in the model" is worth the hassle in the code to satisfy the typing

Reply:
> i THINK (with my limited knowledge of python typing) these where simply not nessecary because the typing is handled by the AnyRepository in models/__init__.py?

With the following code:
```
    def create_repo(
        self,
        repo: AnyRepository,
    ) -> AnyRepositoryResponse:
```

We take input a `AnyRepository` and return a `AnyRepositoryResponse`. Which in theory means that we can have as input a `LocalRepository` and return a `VirtualRepositoryResponse `. Which is not what we want.

The overload function makes sure that `LocalRepository` returns a `LocalRepositoryResponse`.

Reply:
okay i'll revert this and add the duplication again. but this keeps it much more readable imho. I'll squash the intermediate commits in order to keep the history clean